protobuf/avs.proto

```diff
@@ -136,115 +131,173 @@ enum TaskStatus {
+  // When executing a contract we need at least 2 things:
+  // - target contract address
+  // - the message to send to that contract
+  ContractWriteTask = 1;
+  ContractReadTask  = 2;
+
+  // GraphQL query. Note that a graphql can also be query as a HTTP API, but having its own graphql type make writing query easiser without passing all the data in post as parameter.
+  GraphQLDataQueryTask = 3;
 
-  // Next can be empty. In some kind of block, such as branching, the next is
-  // based on branching condition
-  repeated string next = 4;
-
-  // Transfer eth
-  ETHTransfer eth_transfer = 10;
-  // Run one ore more contracts. The call call also be batched with tool like
-  // multicall to wrap many calls
-  ContractExecution contract_execution = 11;
-  // Make call to a graphql endpoint
-  GraphQLDataQuery graphql_data_query = 12;
   // Make call to a HTTP endpoint
-  HTTPAPICall http_data_query = 13;
+  RestAPITask = 4;
+
   // CustomCode allow to run arbitraty JavaScript.
-  CustomCode custom_code = 14;
-  BranchAction branch = 15;
+  BranchTask = 5;
+  FilterTask = 6;
+  CustomCodeTask = 7;
+}
+
+message TaskNode {
+  TaskType node_type = 1;
+  string id   = 2;
+  string name = 3;
+
+  // based on node_type one and only one of these field are set
+  oneof task_body {
+    // Transfer eth require no calldata etc, just a destination address and an eth amount to be sent
+    ETHTransferNode   eth_transfer = 10;
+
+    // Run one ore more contracts. The call call also be batched with tool like
+    // multicall to wrap many calls. in a contract write, we need to generate signature and send as userops.
+    ContractWriteNode contract_write = 11;
+    // read data fron a target contract
+    ContractQueryNode contract_read = 12;
+    // Make call to a graphql endpoint
+    GraphQLQueryNode  graphql_data_query = 13;
+    // Make call to a HTTP endpoint
+    RestAPINode       rest_api = 14;
+    // CustomCode allow to run arbitraty JavaScript.
+    BranchNode        branch = 15;
+    FilterNode        filter = 16;
+    CustomCodeNode    custom_code = 17;
+  }
 }
 
 message Execution {
-  int64 epoch = 1;
+  int64  epoch        = 1;
   string user_op_hash = 2;
-  string error = 3;
+  string error        = 3;
 }
 
 message Task {
-    UUID id = 1;
-    string owner = 2;
-    string smart_wallet_address = 3;
-    TaskTrigger trigger = 4;
-    repeated TaskAction nodes = 5;
+  UUID   id                   = 1;
+  string owner                = 2;
+  string smart_wallet_address = 3;
+
+  // task won't be check before this
+  int64  start_at   = 5;
+  // task won't be run/check after this
+  int64  expired_at = 6;
+  // arbitrary data about this task. has a limit of 255 character
+  string memo = 7;
+
+  int64 completed_at = 8;
+
+  // repeatable means a task can continue to run even after the first execution.
+  // By default, recurring is false, task will only executed once.
+  bool recurring  = 10;
```

Comment:
If it’s a boolean varialbe, I suggest that we name it `is_recurring`, instead of `recurring`. At least in javascript that norm is with a prefix `is`.

Reply:
okay, we have made a decison about this last year. Checking out the automation hub, I think the below set up makes more sense.
![CleanShot 2024-11-14 at 09 31 34@2x](https://github.com/user-attachments/assets/7605dc1b-6684-4de8-b295-541a53938d8b)

maxExeCount: number
0 - good till cancel, equals to is_recurring
1, 2, 3, .... - the actual execution count

The reason being that with `is_recurring: bool` we couldn’t limit the number of execution needed for a reccuring job. 

Slack workflow called it Frequency:
![CleanShot 2024-11-14 at 09 34 54@2x](https://github.com/user-attachments/assets/9048c8c9-fe59-4646-b2f8-1c6013573cd3)

![CleanShot 2024-11-14 at 09 35 09@2x](https://github.com/user-attachments/assets/14af7886-7ce0-449d-911e-3c4804206359)